submodule repository name change not handled gracefully?

When merging past #2582, developers are reporting errors like the following:

```
[31/146] Performing download step for 'lcm'
FAILED: lcm-prefix/src/lcm-stamp/lcm-download 
cd /home/liang/dev/drake-distro && /usr/bin/git submodule update --init --recursive -- externals/lcm && /home/liang/tools/cmake-3.5.2-Linux-x86_64/bin/cmake -E touch /home/liang/dev/drake-build/lcm-prefix/src/lcm-stamp/lcm-download

fatal: reference is not a tree: 74320f0b58216b0425771aaefbe6f6366797bda6
Unable to checkout '74320f0b58216b0425771aaefbe6f6366797bda6' in submodule path 'externals/lcm'
ninja: build stopped: subcommand failed.
```

@liangfok reports that nuking `drake-distro/.git` fixed it for him.  @jwnimmer-tri speculates that the submodules/cmake integration doesn't cope well with changes in the repository URI, which sounds plausible to me.  I'm going to start poking around, but input from CMake experts would be welcome.

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -332,6 +332,10 @@ foreach(proj ${EXTERNAL_PROJECTS})
     # Compute the path to the submodule for this external.
     file(RELATIVE_PATH ${proj}_GIT_SUBMODULE_PATH ${CMAKE_CURRENT_SOURCE_DIR} ${${proj}_SOURCE_DIR})
 
+    # Initialize the submodule configuration now so parallel downloads do not conflict later.
+    execute_process(COMMAND ${GIT_EXECUTABLE} submodule init -- ${${proj}_GIT_SUBMODULE_PATH}
+      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
+
     # Download externals as Git submodules.
     set(${proj}_DOWNLOAD_COMMAND ${GIT_EXECUTABLE} submodule update --init --recursive -- ${${proj}_GIT_SUBMODULE_PATH})
 
@@ -340,10 +344,6 @@ foreach(proj ${EXTERNAL_PROJECTS})
     else()
       set(${proj}_UPDATE_COMMAND "")
     endif()
-
-    # Initialize the submodule configuration now so parallel downloads do not conflict later.
-    execute_process(COMMAND ${GIT_EXECUTABLE} submodule init -- ${${proj}_GIT_SUBMODULE_PATH}
-      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
 
     # now the actual project
     if(${proj}_IS_CMAKE_POD)
@@ -488,6 +488,10 @@ foreach(example IN ITEMS LittleDog)
   if(EXAMPLES_${example_upper})
     message(STATUS "Installation will include extra example: ${example}")
 
+    # Initialize the submodule configuration now so parallel downloads do not conflict later.
+    execute_process(COMMAND ${GIT_EXECUTABLE} submodule init -- drake/examples/${example}
+      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
+
     # Download examples as Git submodules.
     set(${example}_DOWNLOAD_COMMAND ${GIT_EXECUTABLE} submodule update --init --recursive -- drake/examples/${example})
     if(AUTO_UPDATE_EXTERNALS)
@@ -507,10 +511,6 @@ foreach(example IN ITEMS LittleDog)
     drake_forceupdate(download-${example})
     add_dependencies(drake download-${example})
     add_dependencies(download-all download-${example})
-
-    # Initialize the submodule configuration now so parallel downloads do not conflict later.
-    execute_process(COMMAND ${GIT_EXECUTABLE} submodule init -- drake/examples/${example}
-      WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
   endif()
 endforeach()
 

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -341,6 +341,10 @@ foreach(proj ${EXTERNAL_PROJECTS})
 
     if(AUTO_UPDATE_EXTERNALS)
       set(${proj}_UPDATE_COMMAND ${CMAKE_COMMAND} -E chdir ${CMAKE_CURRENT_SOURCE_DIR} ${${proj}_DOWNLOAD_COMMAND})
+
+      # Synchronize the submodule url now so parallel updates do not conflict later.
+      execute_process(COMMAND ${GIT_EXECUTABLE} submodule sync -- ${${proj}_GIT_SUBMODULE_PATH}
+        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
     else()
       set(${proj}_UPDATE_COMMAND "")
     endif()
@@ -496,6 +500,10 @@ foreach(example IN ITEMS LittleDog)
     set(${example}_DOWNLOAD_COMMAND ${GIT_EXECUTABLE} submodule update --init --recursive -- drake/examples/${example})
     if(AUTO_UPDATE_EXTERNALS)
       set(${example}_UPDATE_COMMAND ${CMAKE_COMMAND} -E chdir ${CMAKE_CURRENT_SOURCE_DIR} ${${example}_DOWNLOAD_COMMAND})
+
+      # Synchronize the submodule url now so parallel updates do not conflict later.
+      execute_process(COMMAND ${GIT_EXECUTABLE} submodule sync -- drake/examples/${example}
+        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR})
     else()
       set(${example}_UPDATE_COMMAND "")
     endif()

